README.md
# humidor
arduino-based humidor


i2c_lcd03_lcd05-start/changes.txt

Changes to file lcd03.h


#define LCD_SETCONTRAST 0x1E
#define LCD_SETBRIGHTNESS 0x1F

  void setContrast(uint8_t);
  void setBrightness(uint8_t);







-----------------------

Changes to file lcd03.cpp

void LCD03::setContrast(uint8_t ts) {
  Wire.beginTransmission(_i2c_address);
  Wire.write(REG_COMMAND);
  Wire.write(LCD_SETBRIGHTNESS);
  Wire.write(ts);
  Wire.endTransmission();
}

void LCD03::setBrightness(uint8_t ts) {
  Wire.beginTransmission(_i2c_address);
  Wire.write(REG_COMMAND);
  Wire.write(LCD_SETBRIGHTNESS);
  Wire.write(ts);
  Wire.endTransmission();
}


i2c_lcd03_lcd05-start/lcd03.cpp
/*
  LCD03.cpp - Arduino library for I2C LCD03 display from Robot Electronics
  see http://www.robot-electronics.co.uk/htm/Lcd03tech.htm
  Copyright (c) 2013 Ben Arblaster.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/
#include "LCD03.h"

#include <inttypes.h>
#include "Arduino.h"
#include <Wire.h>

char _i2c_address;

LCD03::LCD03() {
  // Convert 8-bit address from LCD to 7-bit address for Wire
  _i2c_address = I2C_ADDR>>1;
}

LCD03::LCD03(char i2c_address) {
  // Convert 8-bit address from LCD to 7-bit address for Wire
  _i2c_address = i2c_address>>1;
}

void LCD03::begin(uint8_t cols, uint8_t rows) {
  Wire.begin();
  noCursor();
  clear();
}

void LCD03::setCursor(uint8_t pos) {
  pos++;
  Wire.beginTransmission(_i2c_address);
  Wire.write(REG_COMMAND);
  Wire.write(LCD_CURSORPOS);
  Wire.write(pos);
  Wire.endTransmission();
}

void LCD03::setCursor(uint8_t col, uint8_t row) {
  // convert LiquidCrystal col & rows to LCD03
  col++;
  row++;
  Wire.beginTransmission(_i2c_address);
  Wire.write(REG_COMMAND);
  Wire.write(LCD_CURSORPOSXY);
  Wire.write(row);
  Wire.write(col);
  Wire.endTransmission();
}

void LCD03::tabSet(uint8_t ts) {
  Wire.beginTransmission(_i2c_address);
  Wire.write(REG_COMMAND);
  Wire.write(LCD_TABSET);
  Wire.write(ts);
  Wire.endTransmission();
}

void LCD03::setContrast(uint8_t ts) {
  Wire.beginTransmission(_i2c_address);
  Wire.write(REG_COMMAND);
  Wire.write(LCD_SETBRIGHTNESS);
  Wire.write(ts);
  Wire.endTransmission();
}

void LCD03::setBrightness(uint8_t ts) {
  Wire.beginTransmission(_i2c_address);
  Wire.write(REG_COMMAND);
  Wire.write(LCD_SETBRIGHTNESS);
  Wire.write(ts);
  Wire.endTransmission();
}

void LCD03::createChar(uint8_t location, uint8_t charmap[]) {
  // remap custom chars from 0-7 to match LiquidCrystal
  location += LCD_CUSTOMCHAR_BASE;
  Wire.beginTransmission(_i2c_address);
  Wire.write(REG_COMMAND);
  Wire.write(LCD_CUSTOMCHAR);
  Wire.write(location);
  for(int i = 0; i < 8; i++) {
    Wire.write(charmap[i] |= LCD_CUSTOMCHAR_MASK);
  }
  Wire.endTransmission();
}

// get the free buffer capacity in bytes
uint8_t LCD03::bufferFreeBytes() {
  Wire.requestFrom(_i2c_address, 1);
  if(Wire.available()) {
    return Wire.read();
  }
}

uint16_t LCD03::readKeypad () {  
  Wire.requestFrom(_i2c_address, 3);
  uint8_t lowByte;
  uint8_t highByte;
  uint16_t bothBytes;
 
  // Don't care about the first byte
  Wire.read();
  // Read the low and high bytes
  lowByte = Wire.read();
  highByte = Wire.read();

  bothBytes = (lowByte | highByte << 8);

  return bothBytes;
}

// write a single byte in one transmission
size_t LCD03::write(uint8_t value) {
  // remap custom chars to 0-7 to match LiquidCrystal
  if(value < 8) {
    value += LCD_CUSTOMCHAR_BASE;
  }
  send(value);
  // assume write was a success
  return 1;
}

// write multiple bytes in as many transmissions as required
size_t LCD03::write(const uint8_t *buffer, size_t size) {
  Wire.beginTransmission(_i2c_address);
  //wait until the buffer is empty enough
  while(bufferFreeBytes() < BUFFER_LENGTH);
  Wire.write(REG_COMMAND);
  for(int i = 0; i < size; i++) {    
    // we've filled the I2C buffer, flush it and start a new transmission
    if(i != 0 && i % (BUFFER_LENGTH-1) == 0) {
      Wire.endTransmission();
      //wait until the buffer is empty enough
      while(bufferFreeBytes() < BUFFER_LENGTH);
      Wire.beginTransmission(_i2c_address);
      Wire.write(REG_COMMAND);
    }
    Wire.write(buffer[i]);
  }
  Wire.endTransmission();
  // assume write was a success
  return size;
}

// send a single byte command in a single transmission
void LCD03::send(uint8_t command) {
  Wire.beginTransmission(_i2c_address);
  Wire.write(REG_COMMAND);
  Wire.write(command);
  Wire.endTransmission();
}


i2c_lcd03_lcd05-start/lcd03.h
/*
  LCD03.h - Arduino library for I2C LCD03 display from Robot Electronics
  see http://www.robot-electronics.co.uk/htm/Lcd03tech.htm
  Copyright (c) 2013 Ben Arblaster.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/
#ifndef LCD03_h
#define LCD03_h

#include <inttypes.h>
#include <Wire.h>
#include "Print.h"

// default i2c address
#define I2C_ADDR 0xC6

// registers
#define REG_COMMAND 0x00
#define REG_KEYPADLOW 0x01
#define REG_KEYPADHIGH 0x02
#define REG_VERSION 0x03

// commands
#define LCD_NOOP 0x00
#define LCD_CURSORHOME 0x01
#define LCD_CURSORPOS 0x02
#define LCD_CURSORPOSXY 0x03
#define LCD_CURSOROFF 0x04
#define LCD_CURSORON 0x05
#define LCD_CURSORBLINK 0x06
#define LCD_BACKSPACE 0x08
#define LCD_TAB 0x09
#define LCD_CURSORDOWN 0x0A
#define LCD_CURSORUP 0x0B
#define LCD_CLEARDISPLAY 0x0C
#define LCD_LINEFEED 0x0D
#define LCD_CLEARCOLUMN 0x11
#define LCD_TABSET 0x12
#define LCD_BACKLIGHTON 0x13
#define LCD_BACKLIGHTOFF 0x14
#define LCD_SETCONTRAST 0x1E
#define LCD_SETBRIGHTNESS 0x1F
#define LCD_CUSTOMCHAR 0x1B

// custom chars
#define LCD_CUSTOMCHAR_BASE 0x80
#define LCD_CUSTOMCHAR_MASK 0b11100000

// keypad values
#define KEYPAD_1 0x01
#define KEYPAD_2 0x02
#define KEYPAD_3 0x04
#define KEYPAD_4 0x08
#define KEYPAD_5 0x10
#define KEYPAD_6 0x20
#define KEYPAD_7 0x40
#define KEYPAD_8 0x80
#define KEYPAD_9 0x100
#define KEYPAD_STAR 0x200
#define KEYPAD_0 0x400
#define KEYPAD_HASH 0x800

class LCD03 : public Print {
public:
  // Constructors
  LCD03();
  LCD03(char);
  
  // LiquidCrystal compatible functions  
  void begin(uint8_t, uint8_t);
  void clear();
  void home();
  void setCursor(uint8_t);
  void setCursor(uint8_t, uint8_t);
  size_t write(uint8_t);
  size_t write(const uint8_t *buffer, size_t size);
  void cursor();
  void noCursor(); 
  void blink();
  void noBlink();
  void display();
  void noDisplay();
  void createChar(uint8_t, uint8_t[]);
  
  // Unimplemented LiquidCrystal functions
  //void scrollDisplayLeft();
  //void scrollDisplayRight();
  //void autoscroll();
  //void noAutoscroll();
  //void leftToRight();
  //void rightToLeft();

  // LCD03 extras
  void newLine();
  void cursorDown();
  void cursorUp();
  void clearColumn();
  void backspace();
  void tab();
  void tabSet(uint8_t);
  void backlight();
  void noBacklight();
  uint8_t bufferFreeBytes();
  uint16_t readKeypad();
  void setContrast(uint8_t);
  void setBrightness(uint8_t);

private:
  char _i2c_address;
  
  void send(uint8_t);
};

inline void LCD03::clear() {
  send(LCD_CLEARDISPLAY);
}

inline void LCD03::backlight() {
  send(LCD_BACKLIGHTON);
}

inline void LCD03::noBacklight() {
  send(LCD_BACKLIGHTOFF);
}

// provided for compatibility with LiquidCrystal - only backlight is switched
inline void LCD03::display() {
  send(LCD_BACKLIGHTON);
}

// provided for compatibility with LiquidCrystal - only backlight is switched
inline void LCD03::noDisplay() {
  send(LCD_BACKLIGHTOFF);
}

inline void LCD03::home() {
  send(LCD_CURSORHOME);
}

inline void LCD03::noBlink() {
  send(LCD_CURSORON);
}

inline void LCD03::blink() {
  send(LCD_CURSORBLINK);
}

inline void LCD03::noCursor() {
  send(LCD_CURSOROFF);
}

inline void LCD03::cursor() {
  send(LCD_CURSORON);
}

inline void LCD03::newLine() {
  send(LCD_LINEFEED);
}

inline void LCD03::cursorDown() {
  send(LCD_CURSORDOWN);
}

inline void LCD03::cursorUp() {
  send(LCD_CURSORUP);
}

inline void LCD03::clearColumn() {
  send(LCD_CLEARCOLUMN);
}

inline void LCD03::backspace() {
  send(LCD_BACKSPACE);
}

inline void LCD03::tab() {
  send(LCD_TAB);
}

#endif
